core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java

```diff
@@ -0,0 +1,1059 @@
+      List<BlockingQueue<ResultBatch<T>>> intermediaryOutputs = new ArrayList<>(parallelMergeTasks);
+
+      List<? extends List<Sequence<T>>> partitions =
+          Lists.partition(sequences, sequences.size() / parallelMergeTasks);
+
+      for (List<Sequence<T>> partition : partitions) {
+        BlockingQueue<ResultBatch<T>> outputQueue = new ArrayBlockingQueue<>(queueSize);
+        intermediaryOutputs.add(outputQueue);
+        QueuePusher<ResultBatch<T>> pusher = new QueuePusher<>(outputQueue, hasTimeout, timeoutAt);
+
+        List<BatchedResultsCursor<T>> partitionCursors = new ArrayList<>(sequences.size());
+        for (Sequence<T> s : partition) {
+          partitionCursors.add(new YielderBatchedResultsCursor<>(new SequenceBatcher<>(s, batchSize), orderingFn));
+        }
+        PrepareMergeCombineInputsAction<T> blockForInputsAction = new PrepareMergeCombineInputsAction<>(
+            partitionCursors,
+            pusher,
+            orderingFn,
+            combineFn,
+            yieldAfter,
+            batchSize,
+            cancellationGizmo
+        );
+        tasks.add(blockForInputsAction);
+      }
+
+      for (RecursiveAction task : tasks) {
+        getPool().execute(task);
+      }
+
+      QueuePusher<ResultBatch<T>> outputPusher = new QueuePusher<>(out, hasTimeout, timeoutAt);
+      List<BatchedResultsCursor<T>> intermediaryOutputsCursors = new ArrayList<>(intermediaryOutputs.size());
+      for (BlockingQueue<ResultBatch<T>> queue : intermediaryOutputs) {
+        intermediaryOutputsCursors.add(
+            new BlockingQueueuBatchedResultsCursor<>(queue, orderingFn, hasTimeout, timeoutAt)
+        );
+      }
+      PrepareMergeCombineInputsAction<T> finalMergeAction = new PrepareMergeCombineInputsAction<>(
+          intermediaryOutputsCursors,
+          outputPusher,
+          orderingFn,
+          combineFn,
+          yieldAfter,
+          batchSize,
+          cancellationGizmo
+      );
+
+      getPool().execute(finalMergeAction);
+    }
+
+    /**
+     * Computes maximum number of layer 1 parallel merging tasks given available processors and an estimate of current
+     * {@link ForkJoinPool} utilization. A return value of 1 or less indicates that a serial merge will be done on
+     * the pool instead.
+     */
+    int computeNumTasks()
+    {
+      // max is minimum of either number of processors or user suggested parallelism
+      final int maxParallelism = Math.min(JvmUtils.getRuntimeInfo().getAvailableProcessors(), parallelism);
+      // adjust max to be no more than total pool parallelism less the number of running threads + submitted tasks
+      final int utilizationEstimate = getPool().getRunningThreadCount() + getPool().getQueuedSubmissionCount();
+      // minimum of 'max computed parallelism' and pool parallelism less current 'utilization estimate'
+      final int computedParallelism = Math.min(maxParallelism, getPool().getParallelism() - utilizationEstimate);
```

Comment:
Is it possible to return a negative number from `getPool().getParallelism() - utilizationEstimate`? If so, it might be nicer, since that number is logged, to force it to 1 minimum to represent the serial processing case.

Reply:
It is, changed to be at least `1` for friendlier debug logs.